src/Data/UriPath.hs

```diff
@@ -31,31 +34,40 @@ instance s ~ ST => ConvertibleStrings UriPart s where
     convertString = unUriPart
 
 -- | An @UriPath@ is a list of @UriPart@s stored as a difference list.
-newtype UriPath = DiffUriParts { appendUriParts :: [UriPart] -> [UriPart] }
+data UriPath = DiffUriParts
+    { appendUriParts    :: [UriPart] -> [UriPart]
+    , diffUriPartsQuery :: HTTP.Query
+    }
 
 instance Monoid UriPath where
-    mempty = DiffUriParts id
+    mempty = DiffUriParts id []
 
-    DiffUriParts ps `mappend` DiffUriParts qs = DiffUriParts (ps . qs)
+    DiffUriParts ps q  `mappend` DiffUriParts ps' q' = DiffUriParts (ps . ps') (q <> q')
```

Comment:
One can question whether this is line OK, for instance merging `foo?bar` and `goo?gar` gives `foo/goo?bar&gar`.
* on one side this might be surprising and unexpected
* however this is actually the only total behavior I can think of